v2/internal/frontend/desktop/windows/single_instance.go

```diff
@@ -0,0 +1,160 @@
+			// app is already running
+			hwnd := w32.FindWindowW(windows.StringToUTF16Ptr(className), windows.StringToUTF16Ptr(windowName))
+
+			if hwnd != 0 {
+				data := options.SecondInstanceData{
+					Args: os.Args[1:],
+				}
+				serialized, _ := json.Marshal(data)
+
+				SendMessage(hwnd, string(serialized))
+				// exit second instance of app after sending message
+				os.Exit(0)
+			}
+			// if we got any other unknown error we will just start new application instance
+		}
+	} else {
+		createEventTargetWindow(className, windowName)
+	}
+}
+
+func createEventTargetWindow(className string, windowName string) w32.HWND {
+	// callback handler in the event target window
+	wndProc := func(
+		hwnd w32.HWND, msg uint32, wparam w32.WPARAM, lparam w32.LPARAM,
+	) w32.LRESULT {
+		if msg == w32.WM_COPYDATA {
+			ldata := (*COPYDATASTRUCT)(unsafe.Pointer(lparam))
+
+			if ldata.dwData == WMCOPYDATA_SINGLE_INSTANCE_DATA {
+				serialized := uintptrToString(ldata.lpData)
+
+				var secondInstanceData options.SecondInstanceData
+
+				err := json.Unmarshal([]byte(serialized), &secondInstanceData)
+
+				if err == nil {
+					secondInstanceBuffer <- secondInstanceData
+				}
+			}
+
+			return w32.LRESULT(0)
+		}
+
+		return w32.DefWindowProc(hwnd, msg, wparam, lparam)
+	}
+
+	var class w32.WNDCLASSEX
+	class.Size = uint32(unsafe.Sizeof(class))
+	class.Style = 0
+	class.WndProc = syscall.NewCallback(wndProc)
+	class.ClsExtra = 0
+	class.WndExtra = 0
+	class.Instance = w32.GetModuleHandle("")
+	class.Icon = 0
+	class.Cursor = 0
+	class.Background = 0
+	class.MenuName = nil
+	class.ClassName = windows.StringToUTF16Ptr(className)
+	class.IconSm = 0
+
+	w32.RegisterClassEx(&class)
+
+	// create event window that will not be visible for user
+	hwnd := w32.CreateWindowEx(
+		w32.WS_EX_NOACTIVATE|
+			w32.WS_EX_TRANSPARENT|
+			w32.WS_EX_LAYERED|
+			// WS_EX_TOOLWINDOW prevents this window from ever showing up in the taskbar, which
+			// we want to avoid. If you remove this style, this window won't show up in the
+			// taskbar *initially*, but it can show up at some later point. This can sometimes
+			// happen on its own after several hours have passed, although this has proven
+			// difficult to reproduce. Alternatively, it can be manually triggered by killing
+			// `explorer.exe` and then starting the process back up.
+			// It is unclear why the bug is triggered by waiting for several hours.
+			w32.WS_EX_TOOLWINDOW,
+		windows.StringToUTF16Ptr(className),
+		windows.StringToUTF16Ptr(windowName),
+		w32.WS_OVERLAPPED,
+		0,
+		0,
+		0,
+		0,
+		0,
+		0,
+		w32.GetModuleHandle(""),
+		nil,
+	)
+
+	w32.SetWindowLongPtr(
+		hwnd,
+		w32.GWL_STYLE,
+		// The window technically has to be visible to receive WM_PAINT messages (which are used
+		// for delivering events during resizes), but it isn't displayed to the user because of
+		// the LAYERED style.
+		w32.WS_VISIBLE|w32.WS_POPUP,
+	)
+
+	return hwnd
+}
+
+func uintptrToString(cstr uintptr) string {
```

Comment:
Can we use the `x/sys` [UTF16PtrToString](https://pkg.go.dev/golang.org/x/sys/windows#UTF16PtrToString) for this? That should be possible by converting the `uintptr` to a `*uint16`.

Reply:
Updated. Thank you for hint!